<a href="https://colab.research.google.com/github/jithesh9539/Customer_Segmentation_Final/blob/main/Customer_Segmentation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib as mpl
import nltk
import matplotlib.pyplot as plt
import seaborn as sns
import datetime, nltk, warnings
import matplotlib.cm as cm
import itertools
from pathlib import Path
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_samples, silhouette_score
from sklearn import preprocessing, model_selection, metrics, feature_selection
from sklearn.model_selection import GridSearchCV, learning_curve
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix
from sklearn import neighbors, linear_model, svm, tree, ensemble
from wordcloud import WordCloud, STOPWORDS
from sklearn.ensemble import AdaBoostClassifier
from sklearn.decomposition import PCA
from IPython.display import display, HTML
import plotly.graph_objs as go
from plotly.offline import init_notebook_mode,iplot

**Loading Dataset to Pandas DataFrame**

In [2]:
customer_data=pd.read_csv('/content/data.csv',encoding= 'unicode_escape')

In [3]:
customer_data.head()

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

In [4]:
customer_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [5]:
customer_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [6]:
customer_data.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

**As 1/4 th of the entries in CustomerID column is null it wouldn't be efficient enough to impute the values and hence would be deleting the records to get a better accuracy while creating a model.** 

In [7]:
customer_data.shape

(541909, 8)

In [8]:
customer_data.dropna(axis = 0, subset = ['CustomerID'], inplace = True)

In [9]:
customer_data.shape

(406829, 8)

In [10]:
customer_data.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

**Checking for duplicate records in the dataset**

In [11]:
customer_data.duplicated().sum()

5225

**As there are around 5225 duplicate records we are proceeding to drop the same from dataset**

In [12]:
customer_data.drop_duplicates(inplace=True)

In [13]:
customer_data.shape

(401604, 8)

In [14]:
customer_data.head()

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

**Identifying the total number of Products,Customers and Transactions in the dataset**

In [15]:
pd.DataFrame([{'products': len(customer_data['StockCode'].value_counts()),    
               'transactions': len(customer_data['InvoiceNo'].value_counts()),
               'customers': len(customer_data['CustomerID'].value_counts()),  
              }], columns = ['products', 'transactions', 'customers'], index = ['quantity'])

products  transactions  customers
quantity      3684         22190       4372

**Determining the number of products purchased in every transaction**

In [16]:
det =customer_data.groupby(by=['CustomerID', 'InvoiceNo'], as_index=False)['InvoiceDate'].count()
prod_per_trans= det.rename(columns = {'InvoiceDate':'Number of products'})
prod_per_trans[:20].sort_values('CustomerID')

CustomerID InvoiceNo  Number of products
0      12346.0    541431                   1
1      12346.0   C541433                   1
2      12347.0    537626                  31
3      12347.0    542237                  29
4      12347.0    549222                  24
5      12347.0    556201                  18
6      12347.0    562032                  22
7      12347.0    573511                  47
8      12347.0    581180                  11
12     12348.0    568172                   3
11     12348.0    548955                   5
9      12348.0    539318                  17
10     12348.0    541998                   6
13     12349.0    577609                  73
14     12350.0    543037                  17
18     12352.0    546869                   7
15     12352.0    544156                  15
16     12352.0    545323                   5
17     12352.0    545332                   3
19     12352.0    547390                   8

**From the above data can observe that the Invoice Containing C as prefix is a cancelled transaction and also there are transactions wherein few of the customers has purchased one or more products**

In [20]:
temp = customer_data[['CustomerID', 'InvoiceNo', 'Country']].groupby(['CustomerID', 'InvoiceNo', 'Country']).count()
temp = temp.reset_index(drop = False)
countries = temp['Country'].value_counts()
print('Number of Countries in the Dataset: {}'.format(len(countries)))

Number of Countries in the Dataset: 37


In [21]:
data = dict(type='choropleth',
locations = countries.index,
locationmode = 'country names', z = countries,
text = countries.index, colorbar = {'title':'Order nb.'},
colorscale=[[0, 'rgb(224,255,255)'],
            [0.01, 'rgb(166,206,227)'], [0.02, 'rgb(31,120,180)'],
            [0.03, 'rgb(178,223,138)'], [0.05, 'rgb(51,160,44)'],
            [0.10, 'rgb(251,154,153)'], [0.20, 'rgb(255,255,0)'],
            [1, 'rgb(227,26,28)']],    
reversescale = False)
#_______________________
layout = dict(title='Number of orders per country',
geo = dict(showframe = True, projection={'type':'mercator'}))
#______________
choromap = go.Figure(data = [data], layout = layout)
iplot(choromap, validate=False)

**The above graph shows the percentage of orders from the top 10 countries, sorted by the number of orders. This shows that more than 90% of orders are coming from United Kingdom and no other country even makes up 3% of the orders in the data.**

**Counting the number of transactions corresponding to canceled orders:**

In [23]:
prod_per_trans['order_canceled'] = prod_per_trans['InvoiceNo'].apply(lambda x:int('C' in x))
display(prod_per_trans[:100])

CustomerID InvoiceNo  Number of products  order_canceled
0      12346.0    541431                   1               0
1      12346.0   C541433                   1               1
2      12347.0    537626                  31               0
3      12347.0    542237                  29               0
4      12347.0    549222                  24               0
..         ...       ...                 ...             ...
95     12381.0    570725                   1               0
96     12381.0    574694                  16               0
97     12381.0    580638                   7               0
98     12381.0   C565050                   4               1
99     12383.0    539840                  37               0

[100 rows x 4 columns]

**Distribution of Customers by Countries**

In [26]:
customer_country = customer_data[['Country','CustomerID']].drop_duplicates()
customer_country.groupby(['Country'])['CustomerID'].aggregate('count').reset_index(drop=False).sort_values('CustomerID', ascending=False)

Country  CustomerID
35        United Kingdom        3950
14               Germany          95
13                France          87
30                 Spain          31
3                Belgium          25
32           Switzerland          21
26              Portugal          19
18                 Italy          15
12               Finland          12
1                Austria          11
24                Norway          10
23           Netherlands           9
0              Australia           9
9                Denmark           9
6        Channel Islands           9
7                 Cyprus           8
31                Sweden           8
19                 Japan           8
25                Poland           6
33                   USA           4
5                 Canada           4
36           Unspecified           4
17                Israel           4
15                Greece           4
10                  EIRE           3
22                 Malta           2
34  United Arab Emirates           2
2                Bahrain           2
8         Czech Republic           1
21             Lithuania           1
20               Lebanon           1
27                   RSA           1
28          Saudi Arabia           1
29             Singapore           1
16               Iceland           1
4                 Brazil           1
11    European Community           1

In [24]:
a = prod_per_trans['order_canceled'].sum()
b = prod_per_trans.shape[0]
print('Number of orders canceled: {}/{} ({:.2f}%) '.format(a, b, a/b*100))

Number of orders canceled: 3654/22190 (16.47%) 


In [27]:
customer_data.describe()

Quantity      UnitPrice     CustomerID
count  401604.000000  401604.000000  401604.000000
mean       12.183273       3.474064   15281.160818
std       250.283037      69.764035    1714.006089
min    -80995.000000       0.000000   12346.000000
25%         2.000000       1.250000   13939.000000
50%         5.000000       1.950000   15145.000000
75%        12.000000       3.750000   16784.000000
max     80995.000000   38970.000000   18287.000000

As minimum value for the quantity column is -80995 we can observe th

In [28]:
customer_data[(customer_data['Quantity']<0)].head()

InvoiceNo StockCode  ... CustomerID         Country
141   C536379         D  ...    14527.0  United Kingdom
154   C536383    35004C  ...    15311.0  United Kingdom
235   C536391     22556  ...    17548.0  United Kingdom
236   C536391     21984  ...    17548.0  United Kingdom
237   C536391     21983  ...    17548.0  United Kingdom

[5 rows x 8 columns]

In [29]:
customer_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 401604 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    401604 non-null  object 
 1   StockCode    401604 non-null  object 
 2   Description  401604 non-null  object 
 3   Quantity     401604 non-null  int64  
 4   InvoiceDate  401604 non-null  object 
 5   UnitPrice    401604 non-null  float64
 6   CustomerID   401604 non-null  float64
 7   Country      401604 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 27.6+ MB


In [30]:
df_check = customer_data[customer_data['Quantity'] < 0][['CustomerID','Quantity',
                                                   'StockCode','Description','UnitPrice']]
for index, col in  df_check.iterrows():
    if customer_data[(customer_data['CustomerID'] == col[0]) & (customer_data['Quantity'] == -col[1]) 
                & (customer_data['Description'] == col[2])].shape[0] == 0: 
        print(df_check.loc[index])
        print(15*'-'+'>'+' HYPOTHESIS NOT FULFILLED')
        break

CustomerID        14527
Quantity             -1
StockCode             D
Description    Discount
UnitPrice          27.5
Name: 141, dtype: object
---------------> HYPOTHESIS NOT FULFILLED


**We see that the initial hypothesis is not fulfilled because of the existence of a 'Discount' entry. I check again the hypothesis but this time discarding the 'Discount' entries:**

In [31]:

df_check = customer_data[(customer_data['Quantity'] < 0) & (customer_data['Description'] != 'Discount')][
                                 ['CustomerID','Quantity','StockCode',
                                  'Description','UnitPrice']]

for index, col in  df_check.iterrows():
    if customer_data[(customer_data['CustomerID'] == col[0]) & (customer_data['Quantity'] == -col[1]) 
                & (customer_data['Description'] == col[2])].shape[0] == 0: 
        print(index, df_check.loc[index])
        print(15*'-'+'>'+' HYPOTHESIS NOT FULFILLED')
        break

154 CustomerID                               15311
Quantity                                    -1
StockCode                               35004C
Description    SET OF 3 COLOURED  FLYING DUCKS
UnitPrice                                 4.65
Name: 154, dtype: object
---------------> HYPOTHESIS NOT FULFILLED


**Once more, we find that the initial hypothesis is not verified. Hence, cancellations do not necessarily correspond to orders that would have been made beforehand.**

In [35]:
df_cleaned = customer_data.copy(deep = True)
df_cleaned['QuantityCanceled'] = 0

entry_to_remove = []
doubtfull_entry = []

for index, col in  customer_data.iterrows():
    if (col['Quantity'] > 0) or col['Description'] == 'Discount': continue        
    df_test = customer_data[(customer_data['CustomerID'] == col['CustomerID']) &
                         (customer_data['StockCode']  == col['StockCode']) & 
                         (customer_data['InvoiceDate'] < col['InvoiceDate']) & 
                         (customer_data['Quantity']   > 0)].copy()
    
    # Cancelation WITHOUT counterpart
    if (df_test.shape[0] == 0): 
        doubtfull_entry.append(index)
    
    # Cancelation WITH a counterpart
    elif (df_test.shape[0] == 1): 
        index_order = df_test.index[0]
        df_cleaned.loc[index_order, 'QuantityCanceled'] = -col['Quantity']
        entry_to_remove.append(index)        
    
    # If various counterparts exist in orders, then we delete the last one
    elif (df_test.shape[0] > 1): 
        df_test.sort_index(axis=0 ,ascending=False, inplace = True)        
        for ind, val in df_test.iterrows():
            if val['Quantity'] < -col['Quantity']: continue
            df_cleaned.loc[ind, 'QuantityCanceled'] = -col['Quantity']
            entry_to_remove.append(index) 
            break    

In the above function, we checked for two cases:

if there are canceled orders that exists without counterparts
if there's at least one counterpart with the exact same quantity
The index of the corresponding canceled order are respectively kept in the doubtfull_entry and entry_to_remove lists whose sizes are:

**Drop the entries in doubtfull_entry and entry_to_remove**

In [36]:

df_cleaned.drop(entry_to_remove, axis = 0, inplace = True)
df_cleaned.drop(doubtfull_entry, axis = 0, inplace = True)




---
**Check if there are still any canceled orders ** ***italicized text***


In [37]:
remaining_entries = df_cleaned[(df_cleaned['Quantity'] < 0) & (df_cleaned['StockCode'] != 'D')]
print("number of entries to delete: {}".format(remaining_entries.shape[0]))
remaining_entries[:5]

number of entries to delete: 37


InvoiceNo StockCode  ...         Country  QuantityCanceled
27329   C538530     22630  ...  United Kingdom                 0
32870   C539195     22942  ...  United Kingdom                 0
33085   C539221     22941  ...            EIRE                 0
33086   C539221     22942  ...            EIRE                 0
77599   C542744     22700  ...  United Kingdom                 0

[5 rows x 9 columns]

**# Drop the remaining entries**

In [38]:
df_cleaned.drop(remaining_entries.index, axis = 0, inplace = True)

In [40]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392809 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   InvoiceNo         392809 non-null  object 
 1   StockCode         392809 non-null  object 
 2   Description       392809 non-null  object 
 3   Quantity          392809 non-null  int64  
 4   InvoiceDate       392809 non-null  object 
 5   UnitPrice         392809 non-null  float64
 6   CustomerID        392809 non-null  float64
 7   Country           392809 non-null  object 
 8   QuantityCanceled  392809 non-null  int64  
dtypes: float64(2), int64(2), object(5)
memory usage: 30.0+ MB


**Stockcode**